Update customizing-metro.md (#19023)
expo/expo: docs/pages/guides/customizing-metro.md

---
title: Metro bundler

> sidebar_title: Bundling with Metro

Reviewer: we need to update "Customizing Webpack" to use the same naming convention if we change this: https://docs.expo.dev/guides/customizing-webpack/

---

import { Terminal } from '~/ui/components/Snippet';
import { YesIcon, NoIcon } from '~/ui/components/DocIcons';

Expo CLI uses Metro during [`npx expo start`](/workflow/expo-cli#develop) and [`npx expo export`](/workflow/expo-cli#exporting) to bundle your JavaScript code and assets. Metro is built and optimized for React Native, and used for large scale applications like Facebook and Instagram.

## Customizing

> You can customize the Metro bundler by creating a **metro.config.js** file in the root of your project. This file should export a [Metro configuration][metro-config] that extends [`@expo/metro-config`][expo-metro-config].

Reviewer: we should explain here why we import `expo/metro-config` rather than `@expo/metro-config`

Author: ```suggestion
You can customize the Metro bundler by creating a **metro.config.js** file in the root of your project. This file should export a [Metro configuration][metro-config] that extends [`expo/metro-config`][expo-metro-config]. Users should import `expo/metro-config` instead of `@expo/metro-config` to ensure version consistency.
```

```

To learn more, refer to the [`metro.config.js` docs](https://facebook.github.io/metro/docs/configuration).

## Assets

Metro resolves files as either source code or assets. Source code is JavaScript, TypeScript, JSON, and other files that are used by your application. [Asset](/guides/assets) are images, fonts, and other files that should not be transformed by Metro. To accommodate large-scale codebases, Metro requires all extensions for both source code and assets to be explicitly defined before starting the bundler. This is done by adding the `resolver.sourceExts` and `resolver.assetExts` options to the Metro configuration. By default we include the following extensions:

- [`resolver.assetExts`](https://github.com/facebook/metro/blob/main/packages/metro-config/src/defaults/defaults.js#L15)
- [`resolver.sourceExts`](https://github.com/facebook/metro/blob/main/packages/metro-config/src/defaults/defaults.js#L15)

### Adding more file extensions to `assetExts`

The most common customization is including extra asset extensions to Metro. In the **metro.config.js** file, add the file extension (without a leading `.`) to `resolver.assetExts`:

```js
// metro.config.js

> const { getDefaultConfig } = require('@expo/metro-config');

Reviewer: intentional to not use `expo/metro-config`?

Author: ```suggestion
const { getDefaultConfig } = require('expo/metro-config');
```

const config = getDefaultConfig(__dirname);

config.resolver.assetExts.push(
  // Adds support for `.db` files for SQLite databases
  'db'
);

module.exports = config;
```

## Minification

Source code is optimized for readability and debugging. To optimize your application for performance, the source code is automatically minified when [compiling](/workflow/expo-cli#compiling) and [exporting](/workflow/expo-cli#exporting). You can also minify your code during development with `npx expo start --minify`, this is sometimes useful for testing production optimizations.

By default, Metro uses `uglify-es` to minify code. According to [this benchmark](https://github.com/privatenumber/minification-benchmarks) `uglify` generally produces the smallest bundles, and is nearly the _slowest_ minifier. Here are some alternative minifiers you can use with Metro.

### Minification: esbuild

You can use [`esbuild`](https://esbuild.github.io/) to minify exponentially faster than `uglify-es` and `terser`. Just follow this guide: [`metro-minify-esbuild` usage](https://github.com/EvanBacon/metro-minify-esbuild#usage).

### Minification: Terser

> You can use [`terser`](https://github.com/terser/terser) instead of `uglify-es` to mangle and compress your project.

Reviewer: we might want to put uglify first if we are explaining terser in reference to uglify

Author: We explain the uglify part in the main section under "Minification"

First, install terser in your project with `yarn add --dev metro-minify-terser`.

Now set terser with `transformer.minifierPath`, and pass in [`terser` options](https://github.com/terser/terser#compress-options) via `transformer.minifierConfig`.

**metro.config.js**

```js

> const { getDefaultConfig } = require('expo/metro-config');

Reviewer: inconsistent import - `expo/metro-config` here and `@expo/metro-config` above

const config = getDefaultConfig(__dirname);

config.transformer.minifierPath = 'metro-minify-terser';
config.transformer.minifierConfig = {
  // Terser options...
};

module.exports = config;
```

### Minification: Uglify

By default, Metro uses [`uglify-es`](https://github.com/mishoo/UglifyJS) to minify and compress your code. You can customize uglify by passing [options](https://github.com/mishoo/UglifyJS#compress-options) to `transformer.minifierConfig`. For example, if you wanted to remove all console logs from your app in production, you can do the following:

**metro.config.js**

```js
const { getDefaultConfig } = require('expo/metro-config');

const config = getDefaultConfig(__dirname);

// Remove all console logs in production...
config.transformer.minifierConfig.compress.drop_console = true;

module.exports = config;
```

Here are all of the [default Uglify options](https://github.com/facebook/metro/blob/b629f44239bbb3414491755185cf19b5834b4b7a/packages/metro-config/src/defaults/index.js#L94-L111) applied in Metro bundler.

## Source map exploration

> A useful way to debug your source code is by exploring the source maps. You can do this easily in any Expo project using [`react-native-bundle-visualizer`](https://github.com/IjzerenHein/react-native-bundle-visualizer). Just install it with `yarn add -D react-native-bundle-visualizer`, then run `npx react-native-bundle-visualizer`.

Reviewer: ```suggestion
A useful way to debug your source code is by exploring the source maps with [`react-native-bundle-visualizer`](https://github.com/IjzerenHein/react-native-bundle-visualizer). Install it with `yarn add -D react-native-bundle-visualizer`, then run `npx react-native-bundle-visualizer`.
```

> This will show you an interactive breakdown of what makes up your JavaScript bundle. Using this you can find large packages that you may not have intended to bundle in your project. The smaller the bundle, the faster your app will start.

Reviewer 2: ```suggestion
This will show you an interactive breakdown of what makes up your JavaScript bundle. Using this, you can find large packages that you may not have intended to bundle in your project. The smaller the bundle, the faster your app will start.
```

> This will show you an interactive breakdown of what makes up your React bundle. Using this you can find large packages that you may not have wanted bundled in your project. The smaller the bundle, the faster your app will start.

Reviewer: ```suggestion
This will show you an interactive breakdown of what makes up your JavaScript bundle. Using this you can find large packages that you may not have intended to bundle in your project. The smaller the bundle, the faster your app will start.
```
